lambda/snapshots_tool_utils.py

```diff
@@ -362,3 +364,98 @@ def copy_remote(snapshot_identifier, snapshot_object):
+        )
+        _AUTOMATED_BACKUP_LIST = response['DBSnapshots']
+
+    return _AUTOMATED_BACKUP_LIST
+
+
+def copy_or_create_db_snapshot(
+    client,
+    db_instance,
+    snapshot_identifier,
+    snapshot_tags,
+    use_automated_backup=True,
+    backup_interval=24,
+):
+
+    if use_automated_backup is False:
+        logger.info(
+            'creating snapshot out of a running db instance: %s'
+            % db_instance['DBInstanceIdentifier']
+        )
+        snapshot_tags.append(
+            {
+                'Key': 'DBInstanceIdentifier',
+                'Value': db_instance['DBInstanceIdentifier'],
+            }
+        )
+        return client.create_db_snapshot(
+            DBSnapshotIdentifier=snapshot_identifier,
+            DBInstanceIdentifier=db_instance['DBInstanceIdentifier'],
+            Tags=snapshot_tags,
+        )
+
+    # Find the latest automted backup and Copy snapshot out of it
+    all_automated_snapshots = get_all_automated_snapshots(client)
+
+    db_automated_snapshots = list(
+        filter(
+            lambda x: x['DBInstanceIdentifier'] == db_instance['DBInstanceIdentifier'],
+            all_automated_snapshots,
+        )
+    )
+
+    # Raise exception if no automated backup found
+    if len(db_automated_snapshots) <= 0:
+        log_message = (
+            'No automated snapshots found for db: %s'
+            % db_instance['DBInstanceIdentifier']
+        )
+        logger.error(log_message)
+        raise SnapshotToolException(log_message)
+
+    # filter last automated backup
+    db_automated_snapshots.sort(key=lambda x: x['SnapshotCreateTime'])
+    latest_snapshot = db_automated_snapshots[-1]
+
+    # Make sure automated backup is not more than backup_interval window old
+    backup_age = datetime.now(timezone.utc) - latest_snapshot['SnapshotCreateTime']
+    if backup_age.total_seconds() >= (backup_interval * 60 * 60):
+        now = datetime.now()
+        log_message = (
+            'Last automated backup was %s minutes ago. No latest automated backup available. '
+            % ((now - backup_age).total_seconds() / 60)
+        )
+        logger.warn(log_message)
+
+        # If last automated backup is over 2*backup_interval, then raise error
+        if backup_age.total_seconds() >= (backup_interval * 2 * 60 * 60):
+            logger.error(log_message)
+            raise SnapshotToolException(log_message)
+
+    logger.info(
+        'Creating snapshot out of an automated backup: %s'
+        % latest_snapshot['DBSnapshotIdentifier']
+    )
+    snapshot_tags.append(
+        {
+            'Key': 'SourceDBSnapshotIdentifier',
+            'Value': latest_snapshot['DBSnapshotIdentifier'],
+        }
+    )
+
+    return client.copy_db_snapshot(
+        SourceDBSnapshotIdentifier=latest_snapshot['DBSnapshotIdentifier'],
+        TargetDBSnapshotIdentifier=snapshot_identifier,
+        Tags=snapshot_tags,
+        CopyTags=False,
```

Comment:
We should set up an alert for failures in `copy_db_snapshot` following a successful API call. I assume snapshots could still fail from AWS' side after this.

Related:

https://github.com/boto/boto3/issues/960
https://github.com/boto/botocore/issues/1273

Reply:
As discussed , we'll take it up in sprint planning